### PR TITLE
implement utf8_view for replace

### LINKED ISSUE #11913: Update `REPLACE` scalar function to support Utf8View
Part of https://github.com/apache/datafusion/issues/11752 and https://github.com/apache/datafusion/issues/11790

Currently, a call to `REPLACE` with a Utf8View datatypes induces a cast. After the change that fixes this issue, it should not.

```sql
query TT
EXPLAIN SELECT
  REPLACE(column1_utf8view, 'foo', 'bar') as c1,
  REPLACE(column1_utf8view, column2_utf8view, 'bar') as c2
FROM test;
----
logical_plan
01)Projection: replace(__common_expr_1, Utf8("foo"), Utf8("bar")) AS c1, replace(__common_expr_1, CAST(test.column2_utf8view AS Utf8), Utf8("bar")) AS c2
02)--Projection: CAST(test.column1_utf8view AS Utf8) AS __common_expr_1, test.column2_utf8view
03)----TableScan: test projection=[column1_utf8view, column2_utf8view]
```

`REPLACE` is defined here: https://github.com/apache/datafusion/blob/e088945c38b74bb1d86dcbb88a69dfc21d59e375/datafusion/functions/src/string/replace.rs

casting tests are in: https://github.com/apache/datafusion/blob/main/datafusion/sqllogictest/test_files/string_view.slt

### Is your feature request related to a problem or challenge?

We are working to add complete StringView support in DataFusion, which permits potentially much faster processing of string data. See https://github.com/apache/datafusion/issues/10918 for more background.

Today, most DataFusion string functions support DataType::Utf8 and DataType::LargeUtf8 and when called with a StringView argument DataFusion will cast the argument back to DataType::Utf8 which is expensive.

To realize the full speed of StringView, we need to ensure that all string functions support the DataType::Utf8View directly.

### Describe the solution you'd like

Update the function to support DataType::Utf8View directly

### Describe alternatives you've considered

The typical steps are:
1. Write some tests showing the function doesn't support Utf8View (see the tests in [`string_view.slt`](https://github.com/apache/datafusion/blob/main/datafusion/sqllogictest/test_files/string_view.slt) to ensure the arguments are not being cast
2. Change the `Signature` of the function to accept `Utf8View` in addition to `Utf8`/`LargeUtf8`
3. Update the implementation of the function to operate on `Utf8View`

Example PRs
* Update to use an arrow kernel that already supports StringView:  https://github.com/apache/datafusion/pull/11787
* Change the implementation to support StringView directly: https://github.com/apache/datafusion/pull/11676
* Change implementation (option 2): https://github.com/apache/datafusion/pull/11556

### Additional context

The documentation of string functions can be found here: https://datafusion.apache.org/user-guide/sql/scalar_functions.html#string-functions

To test a function with StringView with `datafusion-cli` you can use an example like this (replacing `starts_with` with the relevant function)
```sql
> create table foo as values (arrow_cast('foo', 'Utf8View'), arrow_cast('bar', 'Utf8View'));
0 row(s) fetched.
Elapsed 0.043 seconds.

> select starts_with(column1, column2) from foo;
+--------------------------------------+
| starts_with(foo.column1,foo.column2) |
+--------------------------------------+
| false                                |
+--------------------------------------+
1 row(s) fetched.
Elapsed 0.015 seconds.
```

To see if it is using utf8 view, use `EXPLAIN` to see the plan and verify there is no `CAST`. In this example the `CAST(column1@0 AS Utf8)` indicates that the function is not using `Utf8View` natively

```sql
> explain select starts_with(column1, column2) from foo;
+---------------+------------------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                                         |
+---------------+------------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | Projection: starts_with(CAST(foo.column1 AS Utf8), CAST(foo.column2 AS Utf8))                                                |
|               |   TableScan: foo projection=[column1, column2]                                                                               |
| physical_plan | ProjectionExec: expr=[starts_with(CAST(column1@0 AS Utf8), CAST(column2@1 AS Utf8)) as starts_with(foo.column1,foo.column2)] |
|               |   MemoryExec: partitions=1, partition_sizes=[1]                                                                              |
|               |                                                                                                                              |
+---------------+------------------------------------------------------------------------------------------------------------------------------+
2 row(s) fetched.
Elapsed 0.006 seconds.
```

It is also often good to test with a constant as well (likewise there should be no cast):

```sql
> explain select starts_with(column1, 'foo') from foo;
+---------------+----------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                     |
+---------------+----------------------------------------------------------------------------------------------------------+
| logical_plan  | Projection: starts_with(CAST(foo.column1 AS Utf8), Utf8("foo"))                                          |
|               |   TableScan: foo projection=[column1]                                                                    |
| physical_plan | ProjectionExec: expr=[starts_with(CAST(column1@0 AS Utf8), foo) as starts_with(foo.column1,Utf8("foo"))] |
|               |   MemoryExec: partitions=1, partition_sizes=[1]                                                          |
|               |                                                                                                          |
+---------------+----------------------------------------------------------------------------------------------------------+
2 row(s) fetched.
Elapsed 0.002 seconds.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by thinh2**
## Which issue does this PR close?

Closes #11913 .

## Rationale for this change

Before this change, the `replace` function only support `utf8` and `largeUtf8`.  We used to cast `utf8View` to `utf8`, which reduces the performance of the function. This PR is to add support for `utf8View` inside `replace` function.

## What changes are included in this PR?

- [x] Create a new function to handle `replace` for `utf8View`
- [x] Implement unit test for the `replace` function
- [x] Update `sqllogictest` to match our `utf8View` expectation (no CAST in the logical plans)
- [x] Update the signature of `replace` to handle `utf8View`. Without this change, the sql query will still cast `utf8View` to `utf8`

## Are these changes tested?
- [x] Implement unit test for the `replace` function
- [x] Update `sqllogictest` to match our `utf8View` expectation (no CAST in the logical plans)

## Are there any user-facing changes?
No

### CODE REVIEW

File: datafusion/functions/src/string/replace.rs [hunk 1]
```diff
@@ -66,20 +66,49 @@ impl ScalarUDFImpl for ReplaceFunc {
     }
 
     fn return_type(&self, arg_types: &[DataType]) -> Result<DataType> {
-        utf8_to_str_type(&arg_types[0], "replace")
+        match arg_types[0].clone() {
+            DataType::Utf8 => return Ok(DataType::Utf8),
+            DataType::LargeUtf8 => return Ok(DataType::LargeUtf8),
+            DataType::Utf8View => {
+                println!("utf8view");
```
Comment (alamb) [inline datafusion/functions/src/string/replace.rs:73]: probably left over

File: datafusion/functions/src/string/replace.rs [hunk 2]
```diff
@@ -66,20 +66,49 @@ impl ScalarUDFImpl for ReplaceFunc {
...
             DataType::LargeUtf8 => make_scalar_function(replace::<i64>, vec![])(args),
+            DataType::Utf8View => make_scalar_function(replace_view, vec![])(args),
             other => {
                 exec_err!("Unsupported data type {other:?} for function replace")
             }
         }
     }
 }
 
+fn replace_view(args: &[ArrayRef]) -> Result<ArrayRef> {
```
Comment (alamb) [inline datafusion/functions/src/string/replace.rs:94]: ❤️

File: datafusion/sqllogictest/test_files/string_view.slt [hunk 3]
```diff
@@ -718,9 +718,8 @@ EXPLAIN SELECT
 FROM test;
 ----
 logical_plan
-01)Projection: replace(__common_expr_1, Utf8("foo"), Utf8("bar")) AS c1, replace(__common_expr_1, CAST(test.column2_utf8view AS Utf8), Utf8("bar")) AS c2
-02)--Projection: CAST(test.column1_utf8view AS Utf8) AS __common_expr_1, test.column2_utf8view
-03)----TableScan: test projection=[column1_utf8view, column2_utf8view]
+01)Projection: replace(test.column1_utf8view, Utf8("foo"), Utf8("bar")) AS c1, replace(test.column1_utf8view, test.column2_utf8view, Utf8("bar")) AS c2
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/string_view.slt:721]: I think adding a test that shows this query running (aka run

```sql
SELECT
  REPLACE(column1_utf8view, 'foo', 'bar') as c1,
  REPLACE(column1_utf8view, column2_utf8view, 'bar') as c2
FROM test;
```

in addition to `EXPLAIN`
)

**Review by alamb — COMMENTED** [review]
Looking nice @thinh2  🙏  -- I'll start CI for this PR

File: datafusion/sqllogictest/test_files/string_view.slt (hunk 3 above)
Comment (thinh2) [reply datafusion/sqllogictest/test_files/string_view.slt:721]: Hi @alamb, after updating the replace.rs to process Utf8View, my sql logic test still uses the CAST in its logical plan and this test is failing. Do you have any idea to debug this issue?

Comment (thinh2) [reply datafusion/sqllogictest/test_files/string_view.slt]: Hi @alamb, I already found the issue. The reason is that I didn't update the function's signature to Utf8View.

Comment (tlm365) [reply datafusion/sqllogictest/test_files/string_view.slt]: > after updating the replace.rs to process Utf8View, my sql logic test still uses the CAST in its logical plan and this test is failing. Do you have any idea to debug this issue?

@thinh2 you're missing define `Utf8View` as probably signature [here](https://github.com/apache/datafusion/blob/9f77021c2e5b10c4ee7e98f1c0887524ebf87ae2/datafusion/functions/src/string/replace.rs#L48)

**Review by alamb — APPROVED** [review]
Looks good to me @thinh2  -- thank you 🙏 

I pushed another commit to your branch that adds a slt test to very execution, and merged up from main

Thank you very much for your contribution

File: datafusion/functions/src/string/replace.rs [hunk 4]
```diff
@@ -45,7 +45,11 @@ impl ReplaceFunc {
         use DataType::*;
         Self {
             signature: Signature::one_of(
-                vec![Exact(vec![Utf8, Utf8, Utf8])],
+                vec![
+                    Exact(vec![Utf8View, Utf8View, Utf8View]),
+                    Exact(vec![Utf8, Utf8, Utf8]),
+                    Exact(vec![LargeUtf8, LargeUtf8, LargeUtf8]),
```
Comment (alamb) [inline datafusion/functions/src/string/replace.rs:51]: Thank you for also implementing support for LargeUTF8

**Review by alamb — APPROVED** [review]
Thanks again @thinh2

**Comment by alamb** [comment]
Looks like there are some clippy errors and test updates needed to get CI clean on this PR

**Comment by thinh2** [comment]
Hi @alamb, currently my code fails at this sql logic test https://github.com/apache/datafusion/blob/main/datafusion/sqllogictest/test_files/functions.slt#L825
```
query T
SELECT replace(arrow_cast('foobar', 'Dictionary(Int32, Utf8)'), 'bar', 'hello')
----
foohello
```
The reason is because of replace's signature order. If I push Utf8 before Utf8View, it will pass. However, test in string_view.slt will fail. I am trying to understand the functions test above to fix it.

**Comment by thinh2** [comment]
Hi @alamb, I already fixed all the test errors. Could you please check and run the CI pipeline again?

**Comment by alamb** [comment]
Thank you @thinh2

**Comment by alamb** [comment]
🚀

**Pull request merged**